### PR TITLE
Add support for payload update protocol for generic Titan images.

### PR DESCRIPTION
**Pull request by dawiddabrowski**
1. Add "payload update" as an available command to htool.
2. When "payload update" is invoked, htool will parse the provided image and look for a Titan descriptor, validating the image.
3. If the image is validated, htool will follow the generic "payload update" protocol with the hoth device.
3a. First, htool will send an INITIATE command to the hoth device.
3b. Then, htool will partition the image into chunks of allowed size and send the chunks.
3c. Lastly, htool will finish the protocol by sending a FINALIZE command.

Out of scope:
1. Supporting specific Titan image / hoth device families. Those reside at an offset 52 within the descriptor. Different commands may need to be sent for those devices in the payload update protocol.

### CODE REVIEW

**Comment by gyandreev** [comment]
Please add the new files `htool_payload_update.{c,h}` to the examples/BUILD file

**Comment by dawiddabrowski** [comment]
I fixed the BUILD rule, a #define guard and ran clang-format.

File: examples/htool_payload_update.c [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+  if (fd == -1) {
+    fprintf(stderr, "Error opening file %s: %s\n", image_file, strerror(errno));
+    return -1;
+  }
+  struct stat statbuf;
+  if (fstat(fd, &statbuf)) {
+    fprintf(stderr, "fstat error: %s\n", strerror(errno));
+    return -1;
+  }
+  if (statbuf.st_size > SIZE_MAX) {
```
[the remark this answers is not in the document]
Comment (dawiddabrowski) [reply examples/htool_payload_update.c:129]: I found this check elsewhere in this codebase. It looks like st_size is of a signed type but size_t is an unsigned type and therefore people do this check to make sure that casting st_size into size_t will always work.

File: examples/htool_payload_update.c [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    fprintf(stderr, "file to large\n");
+    return -1;
+  }
+
+  uint8_t *image = mmap(NULL, statbuf.st_size, PROT_READ, MAP_PRIVATE, fd, 0);
+
+  if (!find_image_descriptor(image, statbuf.st_size, TITAN_IMAGE_DESCRIPTOR_ALIGNMENT, TITAN_IMAGE_DESCRIPTOR_MAGIC)) {
+    fprintf(stderr, "Not a valid Titan image.");
```
[the remark this answers is not in the document]
Comment (dawiddabrowski) [reply examples/htool_payload_update.c]: Done.

File: examples/htool_payload_update.c [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+  if (fstat(fd, &statbuf)) {
+    fprintf(stderr, "fstat error: %s\n", strerror(errno));
+    return -1;
+  }
+  if (statbuf.st_size > SIZE_MAX) {
+    fprintf(stderr, "file to large\n");
+    return -1;
+  }
+
+  uint8_t *image = mmap(NULL, statbuf.st_size, PROT_READ, MAP_PRIVATE, fd, 0);
```
[the remark this answers is not in the document]
Comment (dawiddabrowski) [reply examples/htool_payload_update.c:134]: Done.

File: examples/htool_payload_update.c [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+#include "htool.h"
+#include "htool_cmd.h"
+
+const int64_t TITAN_IMAGE_DESCRIPTOR_MAGIC = 0x5F435344474D495F;
+const int64_t TITAN_IMAGE_DESCRIPTOR_ALIGNMENT = 1 << 16;
+
+bool find_image_descriptor(uint8_t *image, size_t len, size_t offset_alignment, int64_t magic) {
+  for (size_t offset = 0; offset < len; offset += offset_alignment) {
+    int64_t magic_candidate;
+    memcpy(&magic_candidate, image + offset, sizeof(int64_t));
```
[the remark this answers is not in the document]
Comment (dawiddabrowski) [reply examples/htool_payload_update.c:39]: Done. Thanks!!!

File: examples/htool_payload_update.c [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+  ret = send_payload_update_request_with_command(dev, PAYLOAD_UPDATE_FINALIZE);
+  if (ret != 0) {
+    fprintf(stderr, "Error when finalizing.");
+    goto cleanup;
+  }
+
+  return 0;
+
+  cleanup:
+  ret = munmap(image, statbuf.st_size);
```
[the remark this answers is not in the document]
Comment (dawiddabrowski) [reply examples/htool_payload_update.c:177]: Done.

File: examples/htool_payload_update.c [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+int send_payload_update_request_with_command(struct libhoth_device *dev,
+                                                    uint8_t command) {
+  struct payload_update_packet request;
+  request.type = command;
+  request.offset = 0;
+  request.len = 0;
+
+  int ret = htool_exec_hostcmd(
+      dev, EC_CMD_BOARD_SPECIFIC_BASE | EC_PRV_CMD_HOTH_PAYLOAD_UPDATE, 0,
```
[the remark this answers is not in the document]
Comment (dawiddabrowski) [reply examples/htool_payload_update.c]: Done.

File: examples/htool_payload_update.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (rkr35) [reply examples/htool_payload_update.c:130]: Ah, I see. Critically, `st_size` is likely a signed _64-bit_ type, and the check ensures the cast down is non-truncating.

**Review by gyandreev — APPROVED** [review]

File: examples/htool_payload_update.c [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    request.offset = offset;
+    request.len = chunk_size;
+    request.type = PAYLOAD_UPDATE_CONTINUE;
+
+    uint8_t buffer[sizeof(struct payload_update_packet) + MAILBOX_SIZE];
+    memcpy(buffer, &request, sizeof(request));
+    memcpy(buffer + sizeof(request), image + offset, chunk_size);
+
+    int ret = htool_exec_hostcmd(
+        dev, EC_CMD_BOARD_SPECIFIC_BASE | EC_PRV_CMD_HOTH_PAYLOAD_UPDATE, 0,
```
Comment (rkr35) [inline examples/htool_payload_update.c:97]: `+` instead of `|`

Comment (dawiddabrowski) [reply examples/htool_payload_update.c]: Done.

**Pull request merged**